keps/sig-node/20190626-node-shutdown-state.md

```diff
@@ -0,0 +1,237 @@
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those
+approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+In case of an instance being shutdown (i.e. not running) the control plane doesn't make the right
+assumptions to enable stateful workload to fail-over, This KEP introduces a flow to do so.
+
+## Motivation
+
+One of the fundamental guarantees of Kubernetes is Pod safety[1], but in some specific situations it’s been too conservative. When a node is shutdown the control plane do not distinguish between a kubelet or node failure, this leads to a situation where it cannot make the right assumptions to preserve the availability of stateful workloads, this manifests as volumes not being detached due the fact that the control plane is not able to determine if containers are still running.
+
+The impact of this behaviour wasn’t visible as when a node is shutdown some of the cloud providers (if not most) were deleting the node object. This led the podgc controller to remove orphaned pods using a 0 grace-period[2]. As a consequence, the attach_detach controller received a watch notification for pod deletion and removed that pod and its attachment from its desired state. Finally when reconciling the actual state with desired state, the reconciler triggers the detach.
+
+This behaviour was very racy, as while the node was being deleted, it could come back and register itself, list its pods and start the containers while volumes are being detached, leading to data corruption. To mitigate this, we explicitly set requirements on handling the node objects for cloud providers[3], i.e. not deleting node object unless the instance do not exist anymore.
+
+This made it noticeable to users, several were complaining about it, as a temporary mitigation we introduced taint called node-shutdown enabling cluster administrators to write automation against it, although most of the solutions at the time were racy.
```

Comment:
Can you be more specific on the current state?  If I read https://github.com/kubernetes/kubernetes/issues/58635 correctly, the node shutdown taint addition was reverted.  When it was last discussed with sig-node, we asked for the taint to be more specific to behavior (e.g. "no-mount" to mirror no-execute), rather than a scenario, "node-shutdown".

Reply:
kubernetes/kubernetes#58635 was reverted due to some issues, but the implementation went it. given that it's a taint on the node object I'd expect we cannot remove it anytime soon due API guarentees. 

defering to @smarterclayton @liggitt

Reply:
It only has API guarantees if it was released with a minor version.  Can you point me to the PR(s) for the "implementation" that went in, and weren't reverted?

Reply:
kubernetes/kubernetes#60009 added it back

Reply:
Thanks for the pointer.  There were no API changes involved in that PR, and the kubelet doesn't use it, so there isn't anything preventing us from changing it if we think that is appropriate.

Reply:
Adding a taint is an API change, just like adding/removing/renaming a label.  It's a slightly different set of constraints, but for instance beta.x.node.kubernetes.io will outlive us all, so thinking about any field a controller sets for the consumption of others as an API is accurate.

Reply:
I see, thanks @smarterclayton.  We can't stop adding the taint to nodes for API reasons.  However, we aren't required to use this particular taint for this KEP, or add any behavior to the kubelet based on the node shutdown taint.  For the purpose of this KEP, we should consider the choice and design of the taints used as in-scope, especially given the existing taint has not been approved in a KEP as far as i'm aware.